pkg/helm/charts.go

```diff
@@ -306,73 +323,120 @@ func GetKotsKinds(helmApp *apptypes.HelmApp) (kotsutil.KotsKinds, error) {
+	secrets, err := clientSet.CoreV1().Secrets(util.PodNamespace).List(context.TODO(), listOpts)
 	if err != nil {
-		return errors.Wrapf(err, "failed to find release secret with options %v", selectorLabels)
+		return kotsKinds, errors.Wrap(err, "failed to list secrets")
 	}
 
 	if len(secrets.Items) != 1 {
-		return errors.Wrapf(err, "expected 1 secret but found %d with options %v", len(secrets.Items), selectorLabels)
+		return kotsKinds, errors.Errorf("expected to mathc 1 secret, but found %d", len(secrets.Items))
 	}
 
-	data, err := HelmReleaseToSecretData(&helmApp.Release)
+	chartSecret := secrets.Items[0]
+	helmApp, err := helmAppFromSecret(&chartSecret)
 	if err != nil {
-		return errors.Wrap(err, "failed to marshal release to helm data")
+		return kotsKinds, errors.Wrap(err, "failed to convert secret to helm app")
 	}
 
-	secret := secrets.Items[0]
-	secret.Data["release"] = data
+	license, err := GetChartLicenseFromSecretOrDownload(helmApp)
+	if err != nil {
+		return kotsKinds, errors.Wrap(err, "failed to get license")
+	}
+	kotsKinds.License = license
 
-	_, err = clientSet.CoreV1().Secrets(helmApp.Namespace).Update(context.TODO(), &secret, metav1.UpdateOptions{})
+	// "Config" object is in the template secret.
+	for _, template := range helmApp.Release.Chart.Templates {
+		if template.Name != "templates/_replicated/secret.yaml" {
+			continue
+		}
+
+		// Can't use client-go decoder here because raw data contains templates.
+		secret := map[string]interface{}{}
+		if err := yaml.Unmarshal(template.Data, &secret); err != nil {
+			return kotsKinds, errors.Wrap(err, "failed to unmarshal secret data")
+		}
+		encodedConfig := util.GetValueFromMapPath(secret, []string{"data", "config"})
+
+		configData, err := util.Base64DecodeInterface(encodedConfig)
+		if err != nil {
+			return kotsKinds, errors.Wrap(err, "failed to base64 decode config from chart templates")
+		}
+		kotsKinds.Config, err = kotsutil.LoadConfigFromBytes(configData)
+		if err != nil {
+			return kotsKinds, errors.Wrap(err, "failed to get config from chart templates")
+		}
+
+		break
+	}
+
+	encodedConfigValues := util.GetValueFromMapPath(helmApp.Release.Config, []string{"replicated", "app", "configValues"})
+	configValuesData, err := util.Base64DecodeInterface(encodedConfigValues)
 	if err != nil {
-		return errors.Wrap(err, "failed to update release secret")
+		return kotsKinds, errors.Wrap(err, "failed to base64 decode config values from chart release")
 	}
 
-	return nil
+	kotsKinds.ConfigValues, err = kotsutil.LoadConfigValuesFromBytes(configValuesData)
+	if err != nil {
+		return kotsKinds, errors.Wrap(err, "failed to get config values from chart values")
+	}
+
+	return kotsKinds, nil
+}
+
+func getKotsKindsFromConfigSecret(secret *corev1.Secret) (kotsutil.KotsKinds, error) {
+	kotsKinds := kotsutil.EmptyKotsKinds()
+
+	licenseData := secret.Data["license"]
+	if len(licenseData) != 0 {
```

Comment:
Shouldn't there be an error returned if `len(licenseData) == 0`?  I would assume `len(licenseData) == 0` would mean the chart isnt managed by us/this process and we likely should not continue.

Reply:
That would probably be ideal.  But right now license is added dynamically by kots.  If release is rolled back, license data will be wiped out.  Maybe we could throw a license download in here if it's missing, and then it would be a food reason to fail here.

Reply:
ah gotchya, thanks